clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/VeniceAdaptiveBlobTransferTrafficThrottler.java

```diff
@@ -0,0 +1,171 @@
+ * (3) If none of the above rule applies, it will increase idle count, if idle count is greater than the pre-defined
+ * threshold, it will try to increase by 20% (still max 200% of the base rate).
+ *
+ * Base rate is initially set to {@link com.linkedin.venice.ConfigKeys#BLOB_TRANSFER_SERVICE_WRITE_LIMIT_BYTES_PER_SEC}
+ * or {@link com.linkedin.venice.ConfigKeys#BLOB_TRANSFER_CLIENT_READ_LIMIT_BYTES_PER_SEC} and will be adjusted based on
+ * input signals dynamically.
+ */
+public class VeniceAdaptiveBlobTransferTrafficThrottler implements VeniceAdaptiveThrottler {
+  private static final Logger LOGGER = LogManager.getLogger(VeniceAdaptiveBlobTransferTrafficThrottler.class);
+  private static final String THROTTLER_NAME_SUFFIX = "TrafficBlobTransferThrottler";
+  private final String throttlerName;
+  private final List<BooleanSupplier> limiterSuppliers = new ArrayList<>();
+  private final List<BooleanSupplier> boosterSuppliers = new ArrayList<>();
+  private final AtomicReference<GlobalChannelTrafficShapingHandler> globalChannelTrafficShapingHandler =
+      new AtomicReference<>();
+  private final long baseRate;
+  private final int signalIdleThreshold;
+  private int currentFactor = 100;
+  private static final int MIN_FACTOR = 20;
+  private static final int MAX_FACTOR = 200;
+  private final int rateUpdatePercentage;
+  private final boolean isWriteThrottler;
+  private int signalIdleCount;
+
+  public VeniceAdaptiveBlobTransferTrafficThrottler(
+      int singleIdleThreshold,
+      long baseRate,
+      int rateUpdatePercentage,
+      boolean isWriteThrottler) {
+    this.throttlerName = (isWriteThrottler ? "Write" : "Read") + THROTTLER_NAME_SUFFIX;
+    this.baseRate = baseRate;
+    this.signalIdleThreshold = singleIdleThreshold;
+    this.isWriteThrottler = isWriteThrottler;
+    this.rateUpdatePercentage = rateUpdatePercentage;
+    LOGGER.info(
+        "Created adaptive throttler with name: {}, base rate: {}, change delta: {}",
+        throttlerName,
+        baseRate,
+        rateUpdatePercentage);
+  }
+
+  @Override
+  public void registerLimiterSignal(BooleanSupplier supplier) {
+    limiterSuppliers.add(supplier);
+  }
+
+  @Override
+  public void registerBoosterSignal(BooleanSupplier supplier) {
+    boosterSuppliers.add(supplier);
+  }
+
+  @Override
+  public void checkSignalAndAdjustThrottler() {
+    if (globalChannelTrafficShapingHandler.get() == null) {
+      LOGGER.info("Related traffic shaping handler does not exist, will not update the throttling number");
+      return;
+    }
+    boolean isSignalIdle = true;
+    boolean hasLimiterRate = false;
+    for (BooleanSupplier supplier: limiterSuppliers) {
+      if (supplier.getAsBoolean()) {
+        hasLimiterRate = true;
+        isSignalIdle = false;
+        signalIdleCount = 0;
+      }
+    }
+    // If any limiter signal is true do not booster the throttler
+    if (hasLimiterRate) {
+      if (currentFactor > MIN_FACTOR) {
```

Comment:
can be written as 
`currentFactor = Math.max(MIN_FACTOR, currentFactor - rateUpdatePercentage);`